### PR TITLE
[FLINK-10958] [table] Add overload support for user defined function

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request adds the overload support for user defined function.*

## Brief change log

  - *add logical in UserDefinedFunctionUtil#getUserDefinedMethod which identifies the methods which are more applicable than others when multiple methods are applicable.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test that validates that overload take effects*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 1]
```diff
@@ -719,19 +750,28 @@ object UserDefinedFunctionUtils {
     * Compares parameter candidate classes with expected classes. If true, the parameters match.
     * Candidate can be null (acts as a wildcard).
     */
+  private def parameterTypeApplicable(candidate: Class[_], expected: Class[_]): Boolean =
+    parameterTypeEquals(candidate, expected) || parameterTypeSubClassOf(candidate, expected)
+
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: It's all right as `parameterTypeSubClassOf` is only used in `parameterTypeApplicable `.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 2]
```diff
@@ -719,19 +750,28 @@ object UserDefinedFunctionUtils {
...
-    (candidate.isArray && expected.isArray &&
-      (candidate.getComponentType == expected.getComponentType ||
-        expected.getComponentType == classOf[Object]))
+    candidate == null ||
+      candidate == expected ||
+      expected.isPrimitive && Primitives.wrap(expected) == candidate ||
+      // time types
+      candidate == classOf[Date] && (expected == classOf[Int] || expected == classOf[JInt])  ||
+      candidate == classOf[Time] && (expected == classOf[Int] || expected == classOf[JInt]) ||
+      candidate == classOf[Timestamp] && (expected == classOf[Long] || expected == classOf[JLong]) ||
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: OK.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/utils/userDefinedScalarFunctions.scala [hunk 3]
```diff
@@ -310,6 +310,16 @@ class Func20 extends ScalarFunction {
   }
 }
 
+object Func21 extends ScalarFunction {
+  def eval(p: People): String = {
+    p.name
+  }
+
+  def eval(p: Student): String = {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/utils/userDefinedScalarFunctions.scala]: If change the Test case `mutable.MutableList[Student]` to `mutable.MutableList[GraduatedStudent]` we can keep this eval method.

[the remark this answers is not in the document]
Comment (dianfu) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/utils/userDefinedScalarFunctions.scala]: Yes, agree. Will update the PR.

**Comment by dianfu** [comment]
@sunjincheng121 thanks a lot for the review. I have updated the PR according to your comments. Could you help to take a look? Thanks a lot.

**Comment by sunjincheng121** [comment]
@dianfu Thanks for the update!
I think the PR is in a good shape. +1 to merged.
Best,
Jincheng

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/CalcITCase.scala [hunk 4]
```diff
@@ -350,4 +350,62 @@ class CalcITCase extends AbstractTestBase {
       "{9=Comment#3}")
     assertEquals(expected.sorted, StreamITCase.testResults.sorted)
   }
+
+  @Test
+  def testOverload(): Unit = {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/CalcITCase.scala]: @twalthr If you are worried about increasing overloading， I agree that only do the validate test.

[the remark this answers is not in the document]
Comment (dianfu) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/CalcITCase.scala]: Make sense to me. I will create a followup PR to move the tests to UserDefinedScalarFunctionTest.

[the remark this answers is not in the document]
Comment (dianfu) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/CalcITCase.scala]: Have created a follow up PR #7179.

**Pull request closed without merging**